File: complete_search_recursion/chessboard_and_queens.cpp
#include<bits/stdc++.h>
using namespace std;
int counter=0;
int column[8];
int diag1[20];
int diag2[20];
int n=4;
int search(int j){
    if(j==n){
        counter+=1;
    }
    else{
        for (int i=0;i<n;i++){
            if (column[i]==1 or diag1[i+j]==1 or diag2[i-j+n-1]==1)
                    continue;
            column[i]=diag1[i+j]=diag2[i-j+n-1]=1;
            search(j+1);
            column[i]=diag1[i+j]=diag2[i-j+n-1]=0;
            
            
        }
    }

}
int main(){
    memset(column,0,sizeof(column));
    memset(diag1,0,sizeof(diag1));
    memset(diag2,0,sizeof(diag2));
    search(0);
    printf("%d",counter);
}

File: complete_search_recursion/chessboard_and_queens.py
n=8
count=0
def search(col:int):
    global n
    if(col==n):
        global count
        count+=1
        return
    else:
        for row in range(0,n):
            if column[row]==1 or diag1[row+col]==1 or diag2[row-col+n-1]==1 or grid[row][col]=='*':
                continue
          
            column[row]=1
            diag1[row+col]=1
            diag2[row-col+n-1]=1
            search(col+1)
            column[row]=0
            diag1[row+col]=0
            diag2[row-col+n-1]=0
            
    
if __name__=='__main__':
    grid=[str(input()) for i in range(n)]
    reserved=[]
    column=[0 for i in range(0,8)]
    diag1=[0 for i in range(0,17)]
    diag2=[0 for i in range(0,17)]

    search(0)
    print(count)